File: raspberry-pi/backend/app/schemas/websocket.py
"""WebSocket event schemas."""

from datetime import datetime
from typing import Literal
from pydantic import BaseModel

from app.schemas.metrics import MetricsResponse

# PlantMetrics WebSocket event schemas
class PlantMetricsPayload(BaseModel):
    """PLANT_METRICS WebSocket event payload."""

    plantId: int
    timestamp: datetime
    metrics: MetricsResponse
    isHealthy: bool


class PlantMetricsMessage(BaseModel):
    """PLANT_METRICS WebSocket message."""

    type: Literal["PLANT_METRICS"] = "PLANT_METRICS"
    payload: PlantMetricsPayload


# ModuleConnectivity WebSocket event schemas
class ModuleConnectivityUpdate(BaseModel):
    """Module connectivity update information."""

    isOnline: bool
    lastSeen: datetime

    
class ModuleConnectivityPayload(BaseModel):
    """MODULE_CONNECTIVITY WebSocket event payload."""

    moduleId: str
    connectivity: ModuleConnectivityUpdate


class ModuleConnectivityMessage(BaseModel):
    """MODULE_CONNECTIVITY WebSocket message."""

    type: Literal["MODULE_CONNECTIVITY"] = "MODULE_CONNECTIVITY"
    payload: ModuleConnectivityPayload


# EntityChange WebSocket event schemas
class EntityChangePayload(BaseModel):
    """ENTITY_CHANGE WebSocket event payload."""

    entity: Literal["plant", "module"]
    action: Literal["create", "update", "delete"]
    id: int | str


class EntityChangeMessage(BaseModel):
    """ENTITY_CHANGE WebSocket message."""

    type: Literal["ENTITY_CHANGE"] = "ENTITY_CHANGE"
    payload: EntityChangePayload

File: raspberry-pi/frontend/app/lib/types/websocket.ts
import type { Metrics } from "./metrics"

// PLANT_METRICS - Sensor data update (Optimistic Update)
export type PlantMetricsPayload = {
  plantId: number
  metrics: Metrics
}

export type PlantMetricsMessage = {
  type: "PLANT_METRICS"
  payload: PlantMetricsPayload
}

// MODULE_CONNECTIVITY - Live module connectivity status (Optimistic Update)
export type ModuleConnectivityPayload = {
  moduleId: string
  connectivity: {
    isOnline: boolean
    lastSeen: string
  }
}

export type ModuleConnectivityMessage = {
  type: "MODULE_CONNECTIVITY"
  payload: ModuleConnectivityPayload
}

// ENTITY_CHANGE - Cache invalidation signal (Refetch)
export type EntityChangePayload = {
  entity: "plant" | "module"
  action: "create" | "update" | "delete"
  id: number | string
}

export type EntityChangeMessage = {
  type: "ENTITY_CHANGE"
  payload: EntityChangePayload
}

// Union of all possible messages
export type IncomingWebSocketMessage =
  | PlantMetricsMessage
  | ModuleConnectivityMessage
  | EntityChangeMessage
  | { type: "PONG" }
